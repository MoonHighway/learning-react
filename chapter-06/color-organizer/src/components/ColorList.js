import PropTypes from 'prop-types'
import Color from './Color'
import '../../stylesheets/ColorList.scss'

const ColorList = ({ colors=[], onRate=f=>f, onRemove=f=>f }) =>
    <div className="color-list">
        {(colors.length === 0) ?
            <p>No Colors Listed. (Add a Color)</p> :
            colors.map(color =>
                <Color key={color.id}
                       {...color}
                       onRate={(rating) => onRate(color.id, rating)}
                       onRemove={() => onRemove(color.id)} />
            )
        }
    </div>

ColorList.propTypes = {
    colors: PropTypes.array,
    onRate: PropTypes.func,
    onRemove: PropTypes.func
}

export default ColorList
