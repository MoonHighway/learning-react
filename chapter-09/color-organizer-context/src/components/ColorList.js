import '../../stylesheets/ColorList.scss'
import PropTypes from 'prop-types'
import Color from './Color'

const ColorList = ({ colors=[] }) =>
    <div className="color-list">
        {(colors.length === 0) ?
            <p>No Colors Listed. (Add a Color)</p> :
            colors.map(color =>
                <Color key={color.id}
                       {...color} />
            )
        }
    </div>

ColorList.propTypes = {
    colors: PropTypes.array
}

export default ColorList
