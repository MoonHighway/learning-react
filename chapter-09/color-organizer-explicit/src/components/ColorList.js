import '../../stylesheets/ColorList.scss'
import PropTypes from 'prop-types'
import Color from './Color'
import { rateColor, removeColor } from '../actions'
import { sortFunction } from '../lib/array-helpers'

const ColorList = ({ store }) => {
    const { colors, sort } = store.getState()
    const sortedColors = [...colors].sort(sortFunction(sort))
    return (
        <div className="color-list">
            {(colors.length === 0) ?
                <p>No Colors Listed. (Add a Color)</p> :
                sortedColors.map(color =>
                    <Color key={color.id}
                        {...color}
                           onRate={(rating) =>
                               store.dispatch(
                                   rateColor(color.id, rating)
                               )
                       }
                           onRemove={() =>
                               store.dispatch(
                                   removeColor(color.id)
                               )
                       }
                    />
                )
            }
        </div>
    )
}

ColorList.propTypes = {
    store: PropTypes.object
}

export default ColorList
