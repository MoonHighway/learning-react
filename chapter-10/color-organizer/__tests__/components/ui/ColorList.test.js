import toJSON from 'enzyme-to-json'
import { compose } from 'redux'
import ColorList from '../../../src/components/ui/ColorList'

const { shallow, mount } = Enzyme

jest.mock('../../../src/components/ui/Color', () =>
    ({rating, onRate=f=>f, onRemove=f=>f}) =>
        <div className="mockColor">
            <button className="rate" onClick={() => onRate(rating)} />
            <button className="remove" onClick={onRemove} />
        </div>
)

describe("<ColorList /> UI Component", () => {

    afterAll(() => jest.resetAllMocks())

    describe("Rendering UI", () => {

        it("Renders Correctly", () =>
            compose(expect,toJSON,shallow)(
                <ColorList colors={_testColors} />
            ).toMatchSnapshot()
        )

        it("Defaults Properties correctly", () =>
            expect(shallow(<ColorList />).find('p').text())
                .toBe('No Colors Listed. (Add a Color)')
        )

        it("Clicking default rate button do not cause Error", () => {
            mount(<ColorList colors={_testColors} />)
                .find('button.rate')
                .first()
                .simulate('click')
        })

        it("Clicking default remove button do not cause Error", () => {
            mount(<ColorList colors={_testColors} />)
                .find('button.remove')
                .first()
                .simulate('click')
        })

    })

    describe("Rating a Color", () => {

        let _rate = jest.fn()

        beforeAll(() =>
            mount(<ColorList colors={_testColors} onRate={_rate} />)
                .find('button.rate')
                .first()
                .simulate('click')
        )

        it("invokes onRate Handler", () =>
            expect(_rate).toBeCalled()
        )

        it("rates the correct color", () =>
            expect(_rate).toBeCalledWith("8658c1d0-9eda-4a90-95e1-8001e8eb6036",4)
        )

    })

    describe("Removing a Color", () => {

        let _remove = jest.fn()

        beforeAll(() =>
            mount(<ColorList colors={_testColors} onRemove={_remove} />)
                .find('button.remove')
                .last()
                .simulate('click')
        )

        it("invokes onRemove Handler", () =>
            expect(_remove).toBeCalled()
        )

        it("removes the correct color", () =>
            expect(_remove).toBeCalledWith("58d9caee-6ea6-4d7b-9984-65b145031979")
        )

    })

})
