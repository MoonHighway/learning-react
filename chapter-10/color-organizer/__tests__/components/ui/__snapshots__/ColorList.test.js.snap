// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ColorList /> UI Component Rendering UI Renders Correctly 1`] = `
<div
  className="color-list"
>
  <Component
    color="#44ef37"
    id="8658c1d0-9eda-4a90-95e1-8001e8eb6036"
    key="8658c1d0-9eda-4a90-95e1-8001e8eb6036"
    onRate={[Function]}
    onRemove={[Function]}
    rating={4}
    timestamp="Sun Apr 10 2016 12:54:19 GMT-0700 (PDT)"
    title="lawn"
  />
  <Component
    color="#0061ff"
    id="f9005b4e-975e-433d-a646-79df172e1dbb"
    key="f9005b4e-975e-433d-a646-79df172e1dbb"
    onRate={[Function]}
    onRemove={[Function]}
    rating={2}
    timestamp="Mon Apr 11 2016 12:54:31 GMT-0700 (PDT)"
    title="ocean blue"
  />
  <Component
    color="#ff4b47"
    id="58d9caee-6ea6-4d7b-9984-65b145031979"
    key="58d9caee-6ea6-4d7b-9984-65b145031979"
    onRate={[Function]}
    onRemove={[Function]}
    rating={0}
    timestamp="Fri Apr 15 2016 12:54:43 GMT-0700 (PDT)"
    title="tomato"
  />
</div>
`;
